src/kubelet/client/client.go

```diff
@@ -0,0 +1,158 @@
+package client
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	"github.com/newrelic/nri-kubernetes/v2/src/client"
+	"github.com/newrelic/nri-kubernetes/v2/src/prometheus"
+)
+
+const (
+	healthzPath             = "/healthz"
+	defaultHTTPKubeletPort  = 10255
+	defaultHTTPSKubeletPort = 10250
+	defaultTimeout          = time.Millisecond * 5000
+)
+
+// Client implements a client for Kubelet, capable of retrieving prometheus metrics from a given endpoint.
+type Client struct {
+	// TODO: Use a non-sdk logger
+	logger        log.Logger
+	doer          client.HTTPDoer
+	endpoint      url.URL
+	apiServerHost string
+}
+
+type OptionFunc func(kc *Client) error
+
+// WithLogger returns an OptionFunc to change the logger from the default noop logger.
+func WithLogger(logger log.Logger) OptionFunc {
+	return func(kc *Client) error {
+		kc.logger = logger
+		return nil
+	}
+}
+
+// New builds a Client using the given options.
+func New(kc kubernetes.Interface, config config.Mock, inClusterConfig *rest.Config, opts ...OptionFunc) (*Client, error) {
+	c := &Client{
+		logger:        log.New(false, io.Discard),
+		apiServerHost: inClusterConfig.Host,
+	}
+
+	for i, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, fmt.Errorf("applying option #%d: %w", i, err)
+		}
+	}
+
+	conn, err := c.setupConnection(kc, tripperWithBearerToken(inClusterConfig.BearerToken), config)
+
+	if err != nil {
+		return nil, fmt.Errorf("connecting to kubelet: %w", err)
+	}
+
+	c.doer = conn.client
+	c.endpoint = conn.url
+
+	return c, nil
+}
+
+func (c *Client) setupConnection(kc kubernetes.Interface, tripperBearerToken http.RoundTripper, config config.Mock) (*connParams, error) {
+	kubeletPort, err := getKubeletPort(kc, config.NodeName)
+	if err != nil {
+		return nil, fmt.Errorf("getting kubelet port: %w", err)
+	}
```

Comment:
Do we know if it's safe to error out if this fails? That is, if K8s _guarantees_ that the port will always be there. Otherwise we'd need to resort to config/default.
@kang-makes

Reply:
Currently the user can set a port ignoring this field.
However, I believe that is "de facto" guaranteed that is populated even if I was not able to find that written somewhere 😕

Reply:
Do you think it's worth implementing a fallback to the default known port? Not sure myself if that would complicate the logic too much.

Reply:
So far it never happen that a single customer had issues with this, I would try to keep the probing logic smaller if possible to avoid hiding such error. By the way it is not a blocker since the user could still specify a port.